pkg/services/datasources/service/datasource.go

```diff
@@ -252,6 +295,123 @@ func (s *Service) AddDataSource(ctx context.Context, cmd *datasources.AddDataSou
 	})
 }
 
+// This will valid validate the instance settings return a version that is safe to be saved
+func (s *Service) prepareInstanceSettings(ctx context.Context, pluginContext backend.PluginContext, settings *backend.DataSourceInstanceSettings) (*backend.DataSourceInstanceSettings, error) {
+	operation := backend.AdmissionRequestCreate
+
+	// First apply global validation rules -- these are required regardless which plugin we are talking to
+	if len(settings.Name) > maxDatasourceNameLen {
+		return nil, datasources.ErrDataSourceNameInvalid.Errorf("max length is %d", maxDatasourceNameLen)
+	}
+
+	if len(settings.URL) > maxDatasourceUrlLen {
+		return nil, datasources.ErrDataSourceURLInvalid.Errorf("max length is %d", maxDatasourceUrlLen)
+	}
+
+	if settings.Type == "" {
+		return settings, nil // NOOP used in tests
+	}
+
+	// Make sure it is a known plugin type
+	p, found := s.pluginStore.Plugin(ctx, settings.Type)
+	if !found {
+		return nil, errutil.BadRequest("datasource.unknownPlugin",
+			errutil.WithPublicMessage(fmt.Sprintf("plugin '%s' not found", settings.Type)))
+	}
+
+	// When the APIVersion is set, the client must also implement ProcessInstanceSettings
+	if p.APIVersion == "" {
+		if settings.APIVersion != "" {
+			return nil, fmt.Errorf("invalid request apiVersion (datasource does not have one configured)")
+		}
+		return settings, nil // NOOP
+	}
+
+	pb, err := backend.DataSourceInstanceSettingsToProtoBytes(settings)
+	if err != nil {
+		return nil, err
+	}
+
+	req := &backend.AdmissionRequest{
+		Operation:     backend.AdmissionRequestCreate,
+		PluginContext: pluginContext,
+		Kind:          settings.GVK(),
+		ObjectBytes:   pb,
+	}
+
+	// Set the old bytes and change the operation
+	if pluginContext.DataSourceInstanceSettings != nil {
+		req.Operation = backend.AdmissionRequestUpdate
+		req.OldObjectBytes, err = backend.DataSourceInstanceSettingsToProtoBytes(settings)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if true {
```

Comment:
this `if true` can be removed?

Reply:
isn't it just here as an example anyway?   Using the block lets me keep the variable the same even though the rsp variable is a different structure 😜   But again honestly none of this matters much so any style you think is fine, I don't really care

Reply:
Yeah, you can use a scope instead
```go
{
  // scope
}
```

See https://medium.com/golangspec/scopes-in-go-a6042bb4298c

But moving this to a function might be cleaner/easier since prepareInstanceSettings currently is 100 lines of code